ibis/expr/datatypes/core.py

```diff
@@ -9,68 +9,49 @@
 from ibis.common.grounds import Concrete, Singleton
 from ibis.common.temporal import IntervalUnit, TimestampUnit
 from ibis.common.validators import Coercible
 
-# TODO(kszucs): we don't support union types yet
-
-dtype = Dispatcher('dtype')
-
-
-@dtype.register(object)
-def dtype_from_object(value, **kwargs) -> DataType:
-    # TODO(kszucs): implement this in a @dtype.register(type) overload once dtype
-    # turned into a singledispatched function because that overload doesn't work
-    # with multipledispatch
-
-    # TODO(kszucs): support Tuple[int, str] and Tuple[int, ...] typehints
-    # in order to support more kinds of typehints follow the implementation of
-    # Validator.from_annotation
-    origin_type = get_origin(value)
-    if origin_type is None:
-        if isinstance(value, type):
-            if issubclass(value, DataType):
-                return value()
-            elif result := _python_dtypes.get(value):
-                return result
-            elif annots := get_type_hints(value):
-                return Struct(toolz.valmap(dtype, annots))
-            elif issubclass(value, bytes):
-                return bytes
-            elif issubclass(value, str):
-                return string
-            elif issubclass(value, Integral):
-                return int64
-            elif issubclass(value, Real):
-                return float64
-            elif value is type(None):
-                return null
-            else:
-                raise TypeError(
-                    f"Cannot construct an ibis datatype from python type `{value!r}`"
-                )
-        else:
-            raise TypeError(
-                f"Cannot construct an ibis datatype from python value `{value!r}`"
-            )
-    elif issubclass(origin_type, Sequence):
-        (value_type,) = map(dtype, get_args(value))
-        return Array(value_type)
-    elif issubclass(origin_type, Mapping):
-        key_type, value_type = map(dtype, get_args(value))
-        return Map(key_type, value_type)
+
+@lazy_singledispatch
+def dtype(value, nullable=True) -> DataType:
+    """Construct an ibis datatype from a python value."""
+    if isinstance(value, DataType):
+        return value
     else:
-        raise TypeError(f'Value {value!r} is not a valid datatype')
+        return DataType.from_typehint(value)
+
+
+@dtype.register(str)
+def from_string(value):
+    return DataType.from_string(value)
+
+
+@dtype.register("numpy.dtype")
+def from_numpy_dtype(value, nullable=True):
+    return DataType.from_numpy(value, nullable)
+
+
+@dtype.register("pandas.core.dtypes.base.ExtensionDtype")
+def from_pandas_extension_dtype(value, nullable=True):
+    return DataType.from_pandas(value, nullable)
+
+
+@dtype.register("pyarrow.lib.DataType")
+def from_pyarrow(value, nullable=True):
+    return DataType.from_pyarrow(value, nullable)
+
+
+# lock the dispatcher to prevent new types from being registered
+del dtype.register
```

Comment:
This seems like it might be overly restrictive, but I also think it's probably for the best for now until there's a request for custom external data types.

That said, we'll need to extend this shortly after this PR lands to support the various pandas nullable dtypes.